ovirt/resource_ovirt_affinity_group.go

```diff
@@ -0,0 +1,425 @@
+	d.Set("vm_enforcing", affinityGroup.MustVmsRule().MustEnforcing())
+	d.Set("vm_positive", affinityGroup.MustVmsRule().MustPositive())
+
+	hosts := affinityGroup.MustHosts().Slice()
+	hostNames := make([]string, len(hosts))
+	for i, h := range hosts {
+		hostNames[i] = h.MustId()
+	}
+	sort.Strings(hostNames)
+	d.Set("host_list", hostNames)
+
+	vms := affinityGroup.MustVms().Slice()
+	vmNames := make([]string, len(vms))
+	for i, v := range vms {
+		vmNames[i] = v.MustId()
+	}
+	sort.Strings(vmNames)
+	d.Set("vm_list", vmNames)
+
+	return nil
+}
+
+func resourceOvirtAffinityGroupUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*ovirtsdk4.Connection)
+
+	group := ovirtsdk4.NewAffinityGroupBuilder()
+	attributeUpdate := false
+
+	if d.HasChange("name") {
+		group.Name(d.Get("name").(string))
+		attributeUpdate = true
+	}
+	if d.HasChange("description") {
+		group.Description(d.Get("description").(string))
+		attributeUpdate = true
+	}
+	if d.HasChange("cluster_id") {
+		group.Cluster(
+			ovirtsdk4.NewClusterBuilder().
+				Id(d.Get("cluster_id").(string)).
+				MustBuild())
+		attributeUpdate = true
+	}
+
+	vmRuleBuilder := ovirtsdk4.NewAffinityRuleBuilder()
+	vmRuleUpdate := false
+	if d.HasChange("vm_positive") {
+		vmRuleBuilder.Positive(d.Get("vm_positive").(bool))
+		vmRuleUpdate = true
+	}
+	if d.HasChange("vm_enforcing") {
+		vmRuleBuilder.Enforcing(d.Get("vm_enforcing").(bool))
+		vmRuleUpdate = true
+	}
+	if d.HasChange("vm_list") {
+		vmRuleBuilder.Enabled(len(d.Get("vm_list").([]interface{})) > 0)
+		vmRuleUpdate = true
+	}
+	if vmRuleUpdate {
+		group.VmsRule(vmRuleBuilder.MustBuild())
+		attributeUpdate = true
+	}
+
+	hostRuleBuilder := ovirtsdk4.NewAffinityRuleBuilder()
+	hostRuleUpdate := false
+	if d.HasChange("host_positive") {
+		hostRuleBuilder.Positive(d.Get("host_positive").(bool))
+		hostRuleUpdate = true
+	}
+	if d.HasChange("host_enforcing") {
+		hostRuleBuilder.Enforcing(d.Get("host_enforcing").(bool))
+		hostRuleUpdate = true
+	}
+	if d.HasChange("host_list") {
+		hostRuleBuilder.Enabled(len(d.Get("host_list").([]interface{})) > 0)
+		hostRuleUpdate = true
+	}
+	if hostRuleUpdate {
+		group.HostsRule(hostRuleBuilder.MustBuild())
+		attributeUpdate = true
+	}
+
+	if attributeUpdate {
+		log.Printf("[DEBUG] Updating %#v", group.MustBuild())
+		_, err := conn.SystemService().
+			ClustersService().
+			ClusterService(d.Get("cluster_id").(string)).
+			AffinityGroupsService().
+			GroupService(d.Id()).
+			Update().
+			Group(group.MustBuild()).
+			Send()
+		if err != nil {
+			log.Printf("[DEBUG] Error updating affinity group (%s): %s", d.Id(), err)
+			return err
+		}
+	}
+
+	if d.HasChange("vm_list") {
+		vmsService := conn.SystemService().
+			ClustersService().
+			ClusterService(d.Get("cluster_id").(string)).
+			AffinityGroupsService().
+			GroupService(d.Id()).
+			VmsService()
+
+		if err := updateVmList(vmsService, d.Get("vm_list").([]interface{})); err != nil {
```

Comment:
this should accept affinityGroupsGroupService

Reply:
I cannot find that object in go-ovirt?

Reply:
Ah, I presume this means be updated to whatever is passed to it, so relates to your changes above (AffinityGroups.Service.GroupService)